### PR TITLE
Project files uploads handler

### PR DESCRIPTION
**Pull request by jacksonargo**
## `POST: /upload`

Implements batch uploads api and deprecates the `/sheets/upload` endpoint.
The uploads api accepts json documents like:
```json

[
  {
    "upload_type": "sheets",
    "content_type": "application/pdf",
    "project": "01-snake-eater",
    "file_name": "Dio_Brando.pdf",
    "file_bytes": "JVBERi0xLjUKJbXtrvsKNCAwIG9iago8PCAvTGVuZ3RoIDUgMCBSCiAgIC9GaWx0ZXIgL0ZsYXRlRGVjb2RlCj4+CnN0cmVhbQp4nDNUMABCXUMgYWFiqGdppJCcy1XIFahQyAWSMLU0BYkrFKUqhCvkcRkqgGBRuoJ+ooFCejFMiZ6hpSVUuxFIaRpXIBcALJUSiwplbmRzdHJlYW0KZW5kb2JqCjUgMCBvYmoKICAgNzQKZW5kb2JqCjMgMCBvYmoKPDwKICAgL0V4dEdTdGF0ZSA8PAogICAgICAvYTAgPDwgL0NBIDEgL2NhIDEgPj4KICAgPj4KPj4KZW5kb2JqCjIgMCBvYmoKPDwgL1R5cGUgL1BhZ2UgJSAxCiAgIC9QYXJlbnQgMSAwIFIKICAgL01lZGlhQm94IFsgMCAwIDU5NS4yIDg0MS45MiBdCiAgIC9Db250ZW50cyA0IDAgUgogICAvR3JvdXAgPDwKICAgICAgL1R5cGUgL0dyb3VwCiAgICAgIC9TIC9UcmFuc3BhcmVuY3kKICAgICAgL0kgdHJ1ZQogICAgICAvQ1MgL0RldmljZVJHQgogICA+PgogICAvUmVzb3VyY2VzIDMgMCBSCj4+CmVuZG9iagoxIDAgb2JqCjw8IC9UeXBlIC9QYWdlcwogICAvS2lkcyBbIDIgMCBSIF0KICAgL0NvdW50IDEKPj4KZW5kb2JqCjYgMCBvYmoKPDwgL1Byb2R1Y2VyIChjYWlybyAxLjE3LjMgKGh0dHBzOi8vY2Fpcm9ncmFwaGljcy5vcmcpKQogICAvQ3JlYXRpb25EYXRlIChEOjIwMjAwMzMwMDAzMDI3LTA3JzAwKQo+PgplbmRvYmoKNyAwIG9iago8PCAvVHlwZSAvQ2F0YWxvZwogICAvUGFnZXMgMSAwIFIKPj4KZW5kb2JqCnhyZWYKMCA4CjAwMDAwMDAwMDAgNjU1MzUgZiAKMDAwMDAwMDQ4MiAwMDAwMCBuIAowMDAwMDAwMjU5IDAwMDAwIG4gCjAwMDAwMDAxODcgMDAwMDAgbiAKMDAwMDAwMDAxNSAwMDAwMCBuIAowMDAwMDAwMTY2IDAwMDAwIG4gCjAwMDAwMDA1NDcgMDAwMDAgbiAKMDAwMDAwMDY2MyAwMDAwMCBuIAp0cmFpbGVyCjw8IC9TaXplIDgKICAgL1Jvb3QgNyAwIFIKICAgL0luZm8gNiAwIFIKPj4Kc3RhcnR4cmVmCjcxNQolJUVPRgo=",
    "sheets_upload": {
      "part_names": [
        "trumpet"
      ],
      "part_numbers": [
        2
      ]
    }
  }
]

```

The sheet data is stored in a slice of `[]sheet.Sheet` structs. The slice is serialized as json and written to `sheets/__data.json` and `sheets/__data.json-Time.Now().RFC3339()`.
Each sheet has a reference to the file key of the actual media.

The serialized sheets data look like this:
```json
[
  {
    "project": "cheese",
    "part_name": "turnip",
    "part_number": 5,
    "file_key": "a7e6ac4d847758f511e3424396ba93eb.pdf"
  }
]
``` 

The api server returns an array of results for each individual upload request. The results are serialized as json and written in the http response.

```json
[
  {
    "file_name": "Dio_Brando.pdf",
    "code": 200
  }
]
```

### CODE REVIEW

File: api_server.go [hunk 1]
```diff
@@ -278,13 +280,138 @@ func (x *ApiServer) SheetsUpload(w http.ResponseWriter, r *http.Request) {
...
+type UploadType string
+
+func (x UploadType) String() string { return string(x) }
+
+const (
+	UploadTypeClickTrack UploadType = "click_track"
+	UploadTypeSheetMusic UploadType = "sheet_music"
+)
+
+type Upload struct {
```
Comment (jacksonargo) [inline api_server.go:296]: Might consider using protobuf here, if we decide to make requests via js.
We can accept both json and protobuf content types, but for _our_ js client, we'll opt for protobuf.

Comment (jacksonargo) [reply api_server.go:296]: We can also use enums for `UploadTypeClickTrack` and `UploadTypeSheetMusic` if we go with protobuf.

File: api_server.go [hunk 2]
```diff
@@ -278,13 +280,138 @@ func (x *ApiServer) SheetsUpload(w http.ResponseWriter, r *http.Request) {
...
+type ClickTrackUpload struct {
+	ContentType string   `json:"content_type"`
+	Project     string   `json:"project"`
+	PartNames   []string `json:"part_names"`
+	PartNumbers []int    `json:"part_numbers"`
+}
+
+type SheetMusicUpload struct {
+	ContentType string   `json:"content_type"`
+	Project     string   `json:"project"`
```
Comment (jacksonargo) [inline api_server.go:319]: Is every upload associated with a project? Probably. So we might want to put this field in the `Upload` struct.

File: api_server.go [hunk 3]
```diff
@@ -278,13 +280,138 @@ func (x *ApiServer) SheetsUpload(w http.ResponseWriter, r *http.Request) {
 		}
 
 	default:
-		http.Error(w, "", http.StatusMethodNotAllowed)
+		methodNotAllowed(w)
+	}
+}
+
+type UploadType string
```
Comment (jacksonargo) [inline api_server.go:287]: Namespacing could get weird here. Might be time to start splitting out into separate packages.

Comment (jacksonargo) [reply api_server.go:287]: esp if we decide on protobuf.

File: api_server.go [hunk 4]
```diff
@@ -278,13 +280,138 @@ func (x *ApiServer) SheetsUpload(w http.ResponseWriter, r *http.Request) {
...
+		results = append(results, status)
+	}
+	json.NewEncoder(w).Encode(&results)
+}
+
+func (x *ApiServer) handleClickTrack(ctx context.Context, upload Upload) UploadStatus {
+	select {
+	case <-ctx.Done():
+		return UploadStatus{
+			FileName: upload.FileName,
```
Comment (jacksonargo) [inline api_server.go:390]: make this a method

File: pkg/api/upload.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+	}
+
+	var documents []Upload
+	if ok := jsonDecode(r.Body, &documents); !ok {
+		badRequest(w, "")
+		return
+	}
+
+	if len(documents) == 0 {
+		return // nothing to do
```
Comment (jacksonargo) [inline pkg/api/upload.go:67]: Probably want to return an error response here

File: pkg/api/upload.go [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+	Project       string `json:"project"`
+	FileName      string `json:"file_name"`
+	FileBytes     []byte `json:"file_bytes"`
+	ContentType   string `json:"content_type"`
+}
+
+type UploadStatus struct {
+	FileName string
+	Code     int
+	Error    string
```
Comment (jacksonargo) [inline pkg/api/upload.go:34]: add json tags here

File: pkg/api/upload.go [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+		results = append(results, status)
+	}
+	json.NewEncoder(w).Encode(&results)
+}
+
+func (x *Server) handleClickTrack(ctx context.Context, upload *Upload) UploadStatus {
+	return UploadStatus{
+		FileName: upload.FileName,
+		Code:     http.StatusNotImplemented,
+		Error:    "lo sentimos, no implementado",
```
Comment (jacksonargo) [inline pkg/api/upload.go:130]: Probably want a builder function for these

File: cmd/vvgo-uploader/main.go [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+func yesNo(reader *bufio.Reader, pre string) bool {
+	yellow.Printf(":: %s [Y/n]? ", pre)
+	answer, _ := reader.ReadString('\n')
+	answer = strings.TrimSpace(strings.ToLower(answer))
+	return answer == "" || answer == "y" || answer == "yes"
+}
+
+func doUpload(flags Flags, uploads []api.Upload) {
```
Comment (jacksonargo) [inline cmd/vvgo-uploader/main.go:158]: this should be in the api pkg

File: pkg/sheet/sheet.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+func (x Sheet) Validate() error {
+	if x.Project == "" {
+		return ErrMissingProject
+	} else if x.PartName == "" {
+		return ErrMissingPartName
+	} else if x.PartNumber == 0 {
+		return ErrMissingPartNumber
+	} else {
+		return nil
+	}
```
Comment (jacksonargo) [inline pkg/sheet/sheet.go:44]: Need to check sheet name is valid from list

File: pkg/api/upload.go [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+	sheets.Sheets
+}
+
+type Upload struct {
+	UploadType    `json:"upload_type"`
+	*ClixUpload   `json:"clix_upload"`
+	*SheetsUpload `json:"sheets_upload"`
+	Project       string `json:"project"`
+	FileName      string `json:"file_name"`
+	FileBytes     []byte `json:"file_bytes"`
```
Comment (0az) [inline pkg/api/upload.go:30]: I still have concerns over this. What happens when we need to go multipart, say, when someone uploads 4k video? Base64 in JSON can't get chunked.

File: pkg/projects/projects.go [hunk 11]
```diff
@@ -0,0 +1,5 @@
+package projects
+
+func Exists(name string) bool {
+	return name == "01-snake-eater"
```
Comment (0az) [inline pkg/projects/projects.go:4]: I don't think this should get encoded into the binary.

File: pkg/sheets/sheets.go [hunk 12]
```diff
@@ -1,66 +1,154 @@
...
 	"fmt"
+	"github.com/virtual-vgo/vvgo/pkg/log"
 	"github.com/virtual-vgo/vvgo/pkg/storage"
-	"net/http"
-	"strconv"
+	"time"
 )
 
-const BucketName = "sheets"
+const DataFile = "__data.json"
```
Comment (0az) [inline pkg/sheets/sheets.go:14]: What happens when we have concurrent writes?

File: pkg/sheets/sheets.go [hunk 13]
```diff
@@ -1,66 +1,154 @@
 package sheets
 
 import (
+	"bytes"
+	"context"
+	"crypto/md5"
```
Comment (0az) [inline pkg/sheets/sheets.go:6]:
```suggestion
	"crypto/sha256"
```

File: pkg/sheets/sheets.go [hunk 14]
```diff
@@ -1,66 +1,154 @@
...
-		Instrument: r.FormValue("instrument"),
-		PartNumber: partNumber,
+func (x Sheets) Store(ctx context.Context, sheets []Sheet, fileBytes []byte) bool {
+	// first, validate all the sheets
+	if ok := validateSheets(sheets); !ok {
+		return false
+	}
+
+	// hash the pdf bytes
+	fileKey := fmt.Sprintf("%x.pdf", md5.Sum(fileBytes))
```
Comment (0az) [inline pkg/sheets/sheets.go:70]:
```suggestion
	fileKey := fmt.Sprintf("%x.pdf", sha256.Sum256(fileBytes))
```

File: pkg/sheets/sheets.go [hunk 15]
```diff
@@ -1,66 +1,154 @@
...
+func (x Sheet) String() string {
+	return fmt.Sprintf("Project: %s Part: %s-%d", x.Project, x.PartName, x.PartNumber)
+}
+
+func (x Sheet) ObjectKey() string {
+	return x.FileKey
+}
+
+func (x Sheet) Link(bucket string) string {
+	return fmt.Sprintf("/download?bucket=%s&object=%s", bucket, x.ObjectKey())
```
Comment (0az) [inline pkg/sheets/sheets.go:144]: Does this allow arbitrary reads?

File: pkg/sheets/sheets_test.go [hunk 16]
```diff
@@ -1,52 +1,193 @@
 package sheets
 
 import (
+	"bytes"
+	"context"
+	"crypto/md5"
```
Comment (0az) [inline pkg/sheets/sheets_test.go:6]:
```suggestion
	"crypto/sha256"
```

File: pkg/sheets/sheets_test.go [hunk 17]
```diff
@@ -1,52 +1,193 @@
...
+			Project:    "cheese",
+			PartName:   "broccoli",
+			PartNumber: 3,
+		}},
+		fileBytes: []byte("pretend i'm a pdf file ;)"),
+	}
+
+	wantOk := true
+	wantNames := []string{
+		fmt.Sprintf("%x.pdf", md5.Sum([]byte("pretend i'm a pdf file ;)"))),
```
Comment (0az) [inline pkg/sheets/sheets_test.go:120]:
```suggestion
		fmt.Sprintf("%x.pdf", sha256.Sum256([]byte("pretend i'm a pdf file ;)"))),
```

File: pkg/sheets/sheets_test.go [hunk 18]
```diff
@@ -1,52 +1,193 @@
...
+		fmt.Sprintf("%s-%s", DataFile, time.Now().UTC().Format(time.RFC3339)),
+		DataFile,
+	}
+	wantObjects := []storage.Object{
+		{ContentType: "application/pdf", Buffer: *bytes.NewBuffer([]byte("pretend i'm a pdf file ;)"))},
+		{ContentType: "application/json", Buffer: *bytes.NewBuffer([]byte(`[{"project":"cheese","part_name":"turnip","part_number":5,"file_key":"0xff"},{"project":"cheese","part_name":"broccoli","part_number":3,"file_key":"a7e6ac4d847758f511e3424396ba93eb.pdf"}]`))},
+		{ContentType: "application/json", Buffer: *bytes.NewBuffer([]byte(`[{"project":"cheese","part_name":"turnip","part_number":5,"file_key":"0xff"},{"project":"cheese","part_name":"broccoli","part_number":3,"file_key":"a7e6ac4d847758f511e3424396ba93eb.pdf"}]`))},
+	}
```
Comment (0az) [inline pkg/sheets/sheets_test.go:128]: Note to self: rehash.

File: pkg/sheets/sheets.go (hunk 13 above)
Comment (jacksonargo) [reply pkg/sheets/sheets.go:6]: md5 hash is good enough for a pdf file name.

File: pkg/sheets/sheets.go (hunk 15 above)
Comment (jacksonargo) [reply pkg/sheets/sheets.go:144]: Access to this api is protected via basic auth. Additionally, the download handler will only provide download links for whitelisted buckets.

File: pkg/projects/projects.go (hunk 11 above)
Comment (jacksonargo) [reply pkg/projects/projects.go:4]: Ya, I didn't want to go too much into the projects api just yet.

File: pkg/sheets/sheets.go (hunk 12 above)
Comment (jacksonargo) [reply pkg/sheets/sheets.go:14]: The server uses a lock in redis to prevent races.

File: pkg/api/upload.go (hunk 10 above)
Comment (jacksonargo) [reply pkg/api/upload.go:30]: You can chunk base64 encoding. You just only encode the x bytes of the pdf, then the next x bytes and so on. Altho, this api doesn't address multi-part uploads just yet.

File: pkg/projects/projects.go (hunk 11 above)
Comment (0az) [reply pkg/projects/projects.go:4]: Fair.

File: pkg/api/upload.go (hunk 10 above)
Comment (0az) [reply pkg/api/upload.go:30]: That seems mildly horrifying, compared to just separating metadata from the actual file.

Comment (jacksonargo) [reply pkg/api/upload.go:30]: I'm not sure what you mean?

Comment (jacksonargo) [reply pkg/api/upload.go:30]: For partial uploads, we'll need to track the parts and put them back together in the correct order, but I don't think that will be too hard.

Comment (0az) [reply pkg/api/upload.go:30]: Still, base64 imposes a 33% overhead.

**Pull request merged**